### ISSUE TITLE
cwm window manager: click --window seems to be intercepted by the root window or the wm?

### ISSUE THREAD

**Issue by jordansissel** [jordansissel/xdotool#397]
When using cwm and sending a click to a specific window id, cwm seems to be interpreting this as a click on the root window... confusing!

Background: On CWM, when you left-click on the root window (the "desktop background" on X11), it brings up a context menu with a list of application windows.

An easy way to reproduce this (for me, at least), is as follows:

```
# Choose the window by clicking on it.
xdotool selectwindow click --window %1 1

# Alternately, from an xterm window (which sets $WINDOWID env)
xdotool click --window $WINDOWID 1
```

This has xdotool let you chose a window and then attempts to send a left click (aka "mouse button 1" in most configurations). When I do this, the CWM root window context menu pops up. 

I tested this on Windows 11 running cwm on Ubuntu using `Xephyr` as my X11 server.

**Comment by jordansissel** [comment]
It gets weirder. If I'm watching a window's events with `xev` then cwm's context menu does not appear.

```
% xterm &

# In the xterm, run `echo $WINDOWID`
# In my case, this prints 12582924
% xev -id 12582924
```

Now, if I do `xdotool click --window 12582924 1` the cwm root window context menu does *not* appear /and also/ the `xev` program prints the correct ButtonPress/ButtonRelease events.

Fascinating.

**Comment by jordansissel** [comment]
I think I got things kind-of-sorted. Event propogation is causing this. The click event  makes its way to the window manager (cwm) which handles the click even though the window isn't the root window.

Running `cwm` in verbose mode, and reproducing the test with `xdotool getactivewindow click --window %1 1`, I can see:

```
debug3: xev_handle_buttonpress: root: 0x2c8 window: 0x800007 subwindow: 0x0
debug3: xev_handle_buttonrelease: root: 0x2c8 window: 0x800007 subwindow: 0x0
```

WIndow 0x800007 is my terminal window.

This seems like a bug in cwm. The cwm docs indicate the default behavior is that left-click on the root window would bring up the window list context menu:

> Menus are recalled by clicking the mouse on the root window:
>
> M1      Show list of currently defined windows.  Selecting an item will warp to that window, unhiding it if necessary.

However, it seems like the way cwm handles certain mouse/key bindings, the context (client or screen) matters, and if it's a screen binding like the window list "menu-window" feature is(?), then it ignores the window parameter.

https://github.com/leahneukirchen/cwm/blob/c55191fdc5a92fd003c5789bbb27dcbb1ff32b1b/xevents.c#L266-L268

Basically, this means that a button event sent to a specific window can also cause the CWM window manager (due to X11 ancestor window event propagation?) to receive a ButtonPress and will process it as if the root window was involved in that button press.

**Comment by jordansissel** [comment]
Definitely questioning how i really know about x11 tonight, hah.

I’m wondering if xdotool is even invoking XSendEvent correctly…

On Wed, Jul 27, 2022, at 10:29 PM, Jordan Sissel wrote:
> 
> 
> I think I got things kind-of-sorted. Event propogation is causing this. The click event makes its way to the window manager (cwm) which handles the click even though the window isn't the root window.
> 
> Running `cwm` in verbose mode, and reproducing the test with `xdotool getactivewindow click --window %1 1`, I can see:
> 
> `debug3: xev_handle_buttonpress: root: 0x2c8 window: 0x800007 subwindow: 0x0
> debug3: xev_handle_buttonrelease: root: 0x2c8 window: 0x800007 subwindow: 0x0
`

**Comment by jordansissel** [comment]
This small patch seems to work. Main changes are based on re-reading the XSendEvent manpage, specifically:

> To determine which clients should receive the specified events, XSendEvent() uses the propagate argument as follows:
> * If event_mask is the empty set, the event is sent to the client that created the destination window. If that client no longer exists, no event is sent.
> * If propagate is False, the event is sent to every client selecting on destination any of the event types in the event_mask argument.
> * If propagate is True and no clients have selected on destination any of the event types in [event-mask](https://tronche.com/gui/x/xlib/window/attributes/event-and-do-not-propagate.html), the destination is replaced with the closest ancestor of destination for which some client has selected a type in event-mask and for which no intervening window has that type in its [do-not-propagate-mask](https://tronche.com/gui/x/xlib/window/attributes/event-and-do-not-propagate.html). If no such window exists or if the window is an ancestor of the focus window and InputFocus was originally specified as the destination, the event is not sent to any clients. Otherwise, the event is reported to every client selecting on the final destination any of the types specified in event_mask.


Based on this, if xdotool sets the `event_mask` to "the empty set" aka "NoEventMask", then the only client receiving the event should be the one that created the window. I'm not exactly sure if `propagate` being True or False matters when the mask is NoEventMask.

```diff
--- a/xdo.c
+++ b/xdo.c
@@ -864,7 +864,7 @@ int _xdo_mousebutton(const xdo_t *xdo, Window window, int button, int is_press)
         case 5: xbpe.state |= Button5MotionMask; break;
       }
     }
-    ret = XSendEvent(xdo->xdpy, window, True, ButtonPressMask, (XEvent *)&xbpe);
+    ret = XSendEvent(xdo->xdpy, window, False, NoEventMask, (XEvent *)&xbpe);
     XFlush(xdo->xdpy);
     free(active_mod);
     return _is_success("XSendEvent(mousedown)", ret == 0, xdo);
```

With the above patch, I'm able to click in my terminal on cwm.

What an odd issue, and I wonder why I've never really seen this before? Or maybe I missed it in the past?
